# Исследование надежности заемщиков

<h1>Оглавление<span class="tocSkip"></span></h1>
<div class="toc"><ul class="toc-item"><li><span><a href="#Общая-информация" data-toc-modified-id="Общая-информация-1"><span class="toc-item-num">1&nbsp;&nbsp;</span>Общая информация</a></span></li><li><span><a href="#Шаг-1.-Обзор-данных." data-toc-modified-id="Шаг-1.-Обзор-данных.-2"><span class="toc-item-num">2&nbsp;&nbsp;</span>Шаг 1. Обзор данных.</a></span></li><li><span><a href="#Шаг-2.-Предобработка-данных." data-toc-modified-id="Шаг-2.-Предобработка-данных.-3"><span class="toc-item-num">3&nbsp;&nbsp;</span>Шаг 2. Предобработка данных.</a></span><ul class="toc-item"><li><span><a href="#Шаг-2.1-Заполнение-пропусков." data-toc-modified-id="Шаг-2.1-Заполнение-пропусков.-3.1"><span class="toc-item-num">3.1&nbsp;&nbsp;</span>Шаг 2.1 Заполнение пропусков.</a></span></li><li><span><a href="#Шаг-2.2-Проверка-данных-на-аномалии-и-исправления." data-toc-modified-id="Шаг-2.2-Проверка-данных-на-аномалии-и-исправления.-3.2"><span class="toc-item-num">3.2&nbsp;&nbsp;</span>Шаг 2.2 Проверка данных на аномалии и исправления.</a></span></li><li><span><a href="#Шаг-2.3.-Изменение-типов-данных." data-toc-modified-id="Шаг-2.3.-Изменение-типов-данных.-3.3"><span class="toc-item-num">3.3&nbsp;&nbsp;</span>Шаг 2.3. Изменение типов данных.</a></span></li><li><span><a href="#Шаг-2.4.-Удаление-дубликатов." data-toc-modified-id="Шаг-2.4.-Удаление-дубликатов.-3.4"><span class="toc-item-num">3.4&nbsp;&nbsp;</span>Шаг 2.4. Удаление дубликатов.</a></span></li><li><span><a href="#Шаг-2.5.-Формирование-дополнительных-датафреймов-словарей,-декомпозиция-исходного-датафрейма." data-toc-modified-id="Шаг-2.5.-Формирование-дополнительных-датафреймов-словарей,-декомпозиция-исходного-датафрейма.-3.5"><span class="toc-item-num">3.5&nbsp;&nbsp;</span>Шаг 2.5. Формирование дополнительных датафреймов словарей, декомпозиция исходного датафрейма.</a></span></li><li><span><a href="#Шаг-2.6.-Категоризация-дохода." data-toc-modified-id="Шаг-2.6.-Категоризация-дохода.-3.6"><span class="toc-item-num">3.6&nbsp;&nbsp;</span>Шаг 2.6. Категоризация дохода.</a></span></li><li><span><a href="#Шаг-2.7.-Категоризация-целей-кредита." data-toc-modified-id="Шаг-2.7.-Категоризация-целей-кредита.-3.7"><span class="toc-item-num">3.7&nbsp;&nbsp;</span>Шаг 2.7. Категоризация целей кредита.</a></span></li></ul></li><li><span><a href="#Ответы-на-вопросы." data-toc-modified-id="Ответы-на-вопросы.-4"><span class="toc-item-num">4&nbsp;&nbsp;</span>Ответы на вопросы.</a></span><ul class="toc-item"><li><span><a href="#Вопрос-1:-Есть-ли-зависимость-между-количеством-детей-и-возвратом-кредита-в-срок?" data-toc-modified-id="Вопрос-1:-Есть-ли-зависимость-между-количеством-детей-и-возвратом-кредита-в-срок?-4.1"><span class="toc-item-num">4.1&nbsp;&nbsp;</span>Вопрос 1: Есть ли зависимость между количеством детей и возвратом кредита в срок?</a></span></li><li><span><a href="#Вопрос-2.-Есть-ли-зависимость-между-семейным-положением-и-возвратом-кредита-в-срок?" data-toc-modified-id="Вопрос-2.-Есть-ли-зависимость-между-семейным-положением-и-возвратом-кредита-в-срок?-4.2"><span class="toc-item-num">4.2&nbsp;&nbsp;</span>Вопрос 2. Есть ли зависимость между семейным положением и возвратом кредита в срок?</a></span></li><li><span><a href="#Вопрос-3.-Есть-ли-зависимость-между-уровнем-дохода-и-возвратом-кредита-в-срок?" data-toc-modified-id="Вопрос-3.-Есть-ли-зависимость-между-уровнем-дохода-и-возвратом-кредита-в-срок?-4.3"><span class="toc-item-num">4.3&nbsp;&nbsp;</span>Вопрос 3. Есть ли зависимость между уровнем дохода и возвратом кредита в срок?</a></span></li><li><span><a href="#Вопрос-4.-Как-разные-цели-кредита-влияют-на-его-возврат-в-срок?" data-toc-modified-id="Вопрос-4.-Как-разные-цели-кредита-влияют-на-его-возврат-в-срок?-4.4"><span class="toc-item-num">4.4&nbsp;&nbsp;</span>Вопрос 4. Как разные цели кредита влияют на его возврат в срок?</a></span></li></ul></li><li><span><a href="#Общий-вывод:" data-toc-modified-id="Общий-вывод:-5"><span class="toc-item-num">5&nbsp;&nbsp;</span>Общий вывод:</a></span></li></ul></div>

## Общая информация

[*Вернуться к оглавлению*](#Оглавление)

**Входные данные**

 Файл `/datasets/data.csv` - статистика о платёжеспособности клиентов.
 
 **Цель:** выяснить, влияет ли семейное положение и количество детей клиента на факт погашения кредита в срок.
 
 **Задачи:**

Шаг 1. Откройте таблицу и изучите общую информацию о данных

Шаг 2. Предобработка данных

Заполнить пропуски, если есть. Описать возможные причины появления пропусков.  
Проанализировать данные на наличие аномалий и описать возможные причины появления таких данных.  
Привести данные к нужному типу.  
Проверить на наличие явных и неявных дубликатов. Привести возможные причины появления дубликатов.  
Создайте два новых датафрейма, в которых:  
каждому уникальному значению из education соответствует уникальное значение education_id — в первом;  
каждому уникальному значению из family_status соответствует уникальное значение family_status_id — во втором.  
На основании диапазонов, указанных ниже, создайте столбец total_income_category с категориями:
- 0–30000 — 'E';
- 30001–50000 — 'D';
- 50001–200000 — 'C';
- 200001–1000000 — 'B';
- 1000001 и выше — 'A'.  

Создайте функцию, которая на основании данных из столбца purpose сформирует новый столбец purpose_category, в который войдут следующие категории:
- 'операции с автомобилем',
- 'операции с недвижимостью',
- 'проведение свадьбы',
- 'получение образования'.

Шаг 3. Ответьте на вопросы

Есть ли зависимость между количеством детей и возвратом кредита в срок?  
Есть ли зависимость между семейным положением и возвратом кредита в срок?  
Есть ли зависимость между уровнем дохода и возвратом кредита в срок?  
Как разные цели кредита влияют на его возврат в срок?  

Шаг 4. Вывод

## Шаг 1. Обзор данных.

[*Вернуться к оглавлению*](#Оглавление)

In [2]:
import pandas as pd
import numpy as np

data = pd.read_csv('data.csv')

data.head()

children  days_employed  dob_years education  education_id  \
0         1   -8437.673028         42    высшее             0   
1         1   -4024.803754         36   среднее             1   
2         0   -5623.422610         33   Среднее             1   
3         3   -4124.747207         32   среднее             1   
4         0  340266.072047         53   среднее             1   

      family_status  family_status_id gender income_type  debt   total_income  \
0   женат / замужем                 0      F   сотрудник     0  253875.639453   
1   женат / замужем                 0      F   сотрудник     0  112080.014102   
2   женат / замужем                 0      M   сотрудник     0  145885.952297   
3   женат / замужем                 0      M   сотрудник     0  267628.550329   
4  гражданский брак                 1      F   пенсионер     0  158616.077870   

                      purpose  
0               покупка жилья  
1     приобретение автомобиля  
2               покупка жилья  
3  дополнительное образование  
4             сыграть свадьбу

In [3]:
data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 21525 entries, 0 to 21524
Data columns (total 12 columns):
 #   Column            Non-Null Count  Dtype  
---  ------            --------------  -----  
 0   children          21525 non-null  int64  
 1   days_employed     19351 non-null  float64
 2   dob_years         21525 non-null  int64  
 3   education         21525 non-null  object 
 4   education_id      21525 non-null  int64  
 5   family_status     21525 non-null  object 
 6   family_status_id  21525 non-null  int64  
 7   gender            21525 non-null  object 
 8   income_type       21525 non-null  object 
 9   debt              21525 non-null  int64  
 10  total_income      19351 non-null  float64
 11  purpose           21525 non-null  object 
dtypes: float64(2), int64(5), object(5)
memory usage: 2.0+ MB


In [4]:
data.columns

Index(['children', 'days_employed', 'dob_years', 'education', 'education_id',
       'family_status', 'family_status_id', 'gender', 'income_type', 'debt',
       'total_income', 'purpose'],
      dtype='object')

Таблица содержит 12 столбцов:
* `children` - количество детей у заемщика
* `days_employed` - стаж в днях
* `dob_years` - возраст
* `education` - образование
* `education_id` - идентификатор образования
* `family_status` - семейное положение
* `family_status_id` - идентификатор семейного положения
* `gender` - пол
* `income_type` - тип занятости
* `debt` - долг
* `total_income` - доход
* `purpose` - цель получения кредита

Имена столбцов составлены стилистически верно. Переименовывать их не требуется.

## Шаг 2. Предобработка данных.

### Шаг 2.1 Заполнение пропусков.

[*Вернуться к оглавлению*](#Оглавление)

Из анализа таблицы `data.info()` столбцы `days_employed` и `total_income`  имеют пропуски.

В данном разделе проанализируем столбец `total_income` и заполним пропуски.

Подсчет количества пропусков:

In [5]:
data.isna().sum()

children               0
days_employed       2174
dob_years              0
education              0
education_id           0
family_status          0
family_status_id       0
gender                 0
income_type            0
debt                   0
total_income        2174
purpose                0
dtype: int64

Подсчет доли пропусков:

In [6]:
round(data.isna().sum() * 100 / len(data), 2)

children             0.0
days_employed       10.1
dob_years            0.0
education            0.0
education_id         0.0
family_status        0.0
family_status_id     0.0
gender               0.0
income_type          0.0
debt                 0.0
total_income        10.1
purpose              0.0
dtype: float64

Проверка того, что NaN присутствует в одной строке одновременно в столбцах `days_employed` и `total_income`:

In [7]:
data.loc[
    (data['days_employed'].isna()) & 
    (data['total_income'].isna())
][['days_employed', 'total_income']].sum()

days_employed    0.0
total_income     0.0
dtype: float64

Исследуем параметры данных в столбце `total_income`:

In [8]:
data['total_income'].describe()

count    1.935100e+04
mean     1.674223e+05
std      1.029716e+05
min      2.066726e+04
25%      1.030532e+05
50%      1.450179e+05
75%      2.034351e+05
max      2.265604e+06
Name: total_income, dtype: float64

В столбце `total_income` аномальных данных не наблюдается. Значения распределены от 0.2066726e+05 до 22.65604e+05. Среднее значение равно 1.651595e+05. Медианное значение равно 1.450179e+05.

Так как пропуски в столбце `total_income` составляют 10.1% данных, не будем исключать эти строки из анализа и заменим пропущенные значения медианным значением с учетом категории заемщика:

In [9]:
data['total_income'] = (
    data.groupby(['income_type'])['total_income']
    .apply(lambda x: x.fillna(x.median()))
)

Проверим заполненные пропуски в таблице и столбце `total_income`:

In [10]:
data.isna().sum()

children               0
days_employed       2174
dob_years              0
education              0
education_id           0
family_status          0
family_status_id       0
gender                 0
income_type            0
debt                   0
total_income           0
purpose                0
dtype: int64

В таблице были пропущены числовые значения в столбцах о стаже работы и доходах.  
Пропуски составляют 10.1% данных.  
Причинами пропусков возможны сокрытие или отсутствие информации о стаже и доходах заемщика.  
Столбец `total_income` был заполнен медианными значениями, т.к. медианное значение менее чувствительно к отклонениям, в случае появления в таблице величин доходов, значительно превышающих доходы других заемщиков.

### Шаг 2.2 Проверка данных на аномалии и исправления.

[*Вернуться к оглавлению*](#Оглавление)

В данном разделе проведем анализ данных в столбце со стажем работы `days_employed` и возрастом `dob_years`.

Исследуем параметры столбца `dob_years`:

In [11]:
data['dob_years'].describe()

count    21525.000000
mean        43.293380
std         12.574584
min          0.000000
25%         33.000000
50%         42.000000
75%         53.000000
max         75.000000
Name: dob_years, dtype: float64

In [12]:
data[data['dob_years'] == 0]['dob_years'].count()

101

Столбец `dob_years` содержит 101 строку с нулевым значением.

Заменим нулевые значения возраста на медианные значения с учетом категории заемщика.

In [13]:
data['dob_years'] = (
    data.groupby(['income_type'])['dob_years'].
    apply(lambda x: x.replace(0, x.median()))
)

In [14]:
data['dob_years'].describe()

count    21525.000000
mean        43.496167
std         12.231538
min         19.000000
25%         34.000000
50%         43.000000
75%         53.000000
max         75.000000
Name: dob_years, dtype: float64

Нулевые значения в столбце с возрастом теперь отсутствуют.

Ранее было обнаружено, что в столбце `days_employed` пропущено 10.1% данных.

Исследуем параметры данных в столбце `days_employed`:

In [15]:
data['days_employed'].describe()

count     19351.000000
mean      63046.497661
std      140827.311974
min      -18388.949901
25%       -2747.423625
50%       -1203.369529
75%        -291.095954
max      401755.400475
Name: days_employed, dtype: float64

В столбце `days_employed` наблюдаются значения разного знака.

Исследуем положительные значения столбца `days_employed`:

In [16]:
days_employed_positive = data.loc[(data['days_employed']>0)]

In [17]:
days_employed_positive['income_type'].value_counts()

пенсионер      3443
безработный       2
Name: income_type, dtype: int64

In [18]:
days_employed_positive['days_employed'].describe()

count      3445.000000
mean     365004.309916
std       21075.016396
min      328728.720605
25%      346639.413916
50%      365213.306266
75%      383246.444219
max      401755.400475
Name: days_employed, dtype: float64

Анализ столбца `days_employed` показал, что для пенсионеров и безработных данные по стажу в среднем равны 365004 дня или 1000 лет, что не может соответствовать действительности.  
Данные по стажу для пенсионеров и безработных указаны неверно. Установить причину по имеющимся данным затруднительно и этот вопрос необходимо адресовать ответственным сотрудникам по заполнению таблиц.

Исследуем отрицательные значения столбца `days_employed`:

In [19]:
days_employed_negative = data.loc[(data['days_employed']<0)]
days_employed_negative['days_employed'].describe()

count    15906.000000
mean     -2353.015932
std       2304.243851
min     -18388.949901
25%      -3157.480084
50%      -1630.019381
75%       -756.371964
max        -24.141633
Name: days_employed, dtype: float64

In [20]:
days_employed_negative['income_type'].value_counts()

сотрудник          10014
компаньон           4577
госслужащий         1312
студент                1
предприниматель        1
в декрете              1
Name: income_type, dtype: int64

По результатам анализа отрицательных значений столбца `days_employed` видно, что эти значения соответствуют заемщикам, имеющим работу. 

Для исправления данных столбца `days_employed` необходимо:  
1) заменить положительные значения на `NaN`  
2) заменить отрицательные значения на положительные  
3) создать категории заемщиков по возрасту  
4) заменить ячейки с `NaN` медианными значениями в зависимости от возраста

In [21]:
# функция drop_positive_nums меняет положительные значения на NaN
def drop_positive_nums(value):
    if value > 0:
        value = np.NaN
        return value
    else:
        return value

# перезаписываем все положительные на NaN в столбце days_employed
data['days_employed'] = data['days_employed'].apply(drop_positive_nums)

# меняем все отрицательные на положительные в столбце days_employed
data['days_employed'] = abs(data['days_employed'])

# категоризация заемщиков по возрасту
cut_labels = ['18-29', '30-39', '40-49', '50-59', '60-69', '>=70']
cut_bins = [0, 29, 39, 49, 59, 69, 100]
data['dob_group'] = pd.cut(data['dob_years'],
                             bins=cut_bins,
                             labels=cut_labels)

# заменим все NaN на медианные в столбце days_employed
data['days_employed'] = (
    data.groupby(['dob_group'])['days_employed']
    .apply(lambda x: x.fillna(x.median()))
)

Проверим заполненные пропуски в таблице `data`:

In [22]:
data.isna().sum()

children            0
days_employed       0
dob_years           0
education           0
education_id        0
family_status       0
family_status_id    0
gender              0
income_type         0
debt                0
total_income        0
purpose             0
dob_group           0
dtype: int64

В столбце `dob_years` были обнаружены строки с нулевым возрастом. Нулевые значения были заменены на медианные значения в зависисмости от категории заемщика.

В столбце `days_employed` были выявлены аномальные значения стажа для заемщиков:  
1) для пенсионеров и неработающих были указаны большие значения стажа, невозможные в действительности  
2) для работающих стаж указан отрицательными значениями  
Причины аномальных данных не выглядят случайными, т.к. очевидна зависимость данных по стажу от категории заемщиков. Выяснить особенности заполнения данных по стажу возможно обратившись к специалистам, ответственным за заполнение таблиц, или изучив документацию к таблицам.

Для исправления данных в столбце `days_employed` данные для пенсионеров и неработающих были заменены на `NaN`. Отрицательные значения стажа заменены положительными значениями. Все отсутствующие значения заменены медианными величинами в зависимости от возраста заемщика.

### Шаг 2.3. Изменение типов данных.

[*Вернуться к оглавлению*](#Оглавление)

Заменим вещественный тип данных в столбцах `days_employed` и `total_income` на целочисленный:

In [23]:
data['days_employed'] = data[['days_employed']].astype('int')
data['total_income'] = data[['total_income']].astype('int')
data.head()

children  days_employed  dob_years education  education_id  \
0         1           8437         42    высшее             0   
1         1           4024         36   среднее             1   
2         0           5623         33   Среднее             1   
3         3           4124         32   среднее             1   
4         0           2261         53   среднее             1   

      family_status  family_status_id gender income_type  debt  total_income  \
0   женат / замужем                 0      F   сотрудник     0        253875   
1   женат / замужем                 0      F   сотрудник     0        112080   
2   женат / замужем                 0      M   сотрудник     0        145885   
3   женат / замужем                 0      M   сотрудник     0        267628   
4  гражданский брак                 1      F   пенсионер     0        158616   

                      purpose dob_group  
0               покупка жилья     40-49  
1     приобретение автомобиля     30-39  
2               покупка жилья     30-39  
3  дополнительное образование     30-39  
4             сыграть свадьбу     50-59

### Шаг 2.4. Удаление дубликатов.

[*Вернуться к оглавлению*](#Оглавление)

Приведем значения столбцов `education`, `family_status`, `income_type`, `purpose` к нижнему регистру:

In [24]:
data['education'] = data['education'].str.lower()
data['family_status'] = data['family_status'].str.lower()
data['income_type'] = data['income_type'].str.lower()
data['purpose'] = data['purpose'].str.lower()

Проверим данные в столбце `education`:

In [25]:
data['education'].value_counts()

среднее                15233
высшее                  5260
неоконченное высшее      744
начальное                282
ученая степень             6
Name: education, dtype: int64

Проверим данные в столбце `family_status`:

In [26]:
data['family_status'].value_counts()

женат / замужем          12380
гражданский брак          4177
не женат / не замужем     2813
в разводе                 1195
вдовец / вдова             960
Name: family_status, dtype: int64

Проверим данные в столбце `income_type`:

In [27]:
data['income_type'].value_counts()

сотрудник          11119
компаньон           5085
пенсионер           3856
госслужащий         1459
безработный            2
предприниматель        2
студент                1
в декрете              1
Name: income_type, dtype: int64

Проверим данные в столбце `purpose`:

In [28]:
data['purpose'].value_counts()

свадьба                                   797
на проведение свадьбы                     777
сыграть свадьбу                           774
операции с недвижимостью                  676
покупка коммерческой недвижимости         664
покупка жилья для сдачи                   653
операции с жильем                         653
операции с коммерческой недвижимостью     651
покупка жилья                             647
жилье                                     647
покупка жилья для семьи                   641
строительство собственной недвижимости    635
недвижимость                              634
операции со своей недвижимостью           630
строительство жилой недвижимости          626
покупка недвижимости                      624
строительство недвижимости                620
покупка своего жилья                      620
ремонт жилью                              612
покупка жилой недвижимости                607
на покупку своего автомобиля              505
заняться высшим образованием      

Проверим данные в столбце `gender`:

In [29]:
data['gender'].value_counts()

F      14236
M       7288
XNA        1
Name: gender, dtype: int64

In [30]:
data[data['gender'] == 'XNA']

children  days_employed  dob_years            education  education_id  \
10701         0           2358         24  неоконченное высшее             2   

          family_status  family_status_id gender income_type  debt  \
10701  гражданский брак                 1    XNA   компаньон     0   

       total_income               purpose dob_group  
10701        203905  покупка недвижимости     18-29

В таблице была обнаружена одна строка с неизвестным полом. При внесении данных была допущена ошибка.  

Проверим данные в столбце `children`:

In [31]:
data['children'].value_counts()

 0     14149
 1      4818
 2      2055
 3       330
 20       76
-1        47
 4        41
 5         9
Name: children, dtype: int64

В таблице обнаружены ошибки по количеству детей: -1 и 20.  
Заменим значения -1 на 1, т.к., вероятно, была допущена опечатка при внесении данных.
Заменим 20 на медианную величину, т.к. неизвестно какое значение должно было быть 2 или 0.

In [32]:
data['children'] = data['children'].replace(-1, 1)
data['children'] = data['children'].replace(20, np.NaN)
data['children'] = data['children'].fillna(data['children'].median())
data['children'].value_counts()

0.0    14225
1.0     4865
2.0     2055
3.0      330
4.0       41
5.0        9
Name: children, dtype: int64

Удалим дубликаты в таблице:

In [33]:
data = data.drop_duplicates().reset_index(drop=True) 
data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 21454 entries, 0 to 21453
Data columns (total 13 columns):
 #   Column            Non-Null Count  Dtype   
---  ------            --------------  -----   
 0   children          21454 non-null  float64 
 1   days_employed     21454 non-null  int32   
 2   dob_years         21454 non-null  int64   
 3   education         21454 non-null  object  
 4   education_id      21454 non-null  int64   
 5   family_status     21454 non-null  object  
 6   family_status_id  21454 non-null  int64   
 7   gender            21454 non-null  object  
 8   income_type       21454 non-null  object  
 9   debt              21454 non-null  int64   
 10  total_income      21454 non-null  int32   
 11  purpose           21454 non-null  object  
 12  dob_group         21454 non-null  category
dtypes: category(1), float64(1), int32(2), int64(4), object(5)
memory usage: 1.8+ MB


Для поиска дубликатов были проделаны следующие операции:  
* приведение значений строковых переменных к нижнему регистру
* проверка столбцов на уникальность данных  

В столбце с полом заемщика `gender` была обнаружена строка с ошибочным значением `XNA`. Т.к. по имеющимся данным восстановить пол невозможно, значение было оставлено без изменений.  
В столбце с данными о количестве детей были обнаружены ошибочные данные -1 и 20. -1 было заменено на 1, т.к. при внесении данных минус мог быть указан ошибочно. 20 было заменено на медианное значение, т.к. установить исходное верное значение невозможно.  
Для удаления дубликатов был использован метод `drop_duplicates()` с обновлением индексов.

### Шаг 2.5. Формирование дополнительных датафреймов словарей, декомпозиция исходного датафрейма.

[*Вернуться к оглавлению*](#Оглавление)

In [34]:
# Формируем словарь education_id - education
education_dict = data[['education_id', 'education']]
education_dict = education_dict.drop_duplicates().reset_index(drop=True).sort_values(by='education_id')
education_dict

education_id            education
0             0               высшее
1             1              среднее
2             2  неоконченное высшее
3             3            начальное
4             4       ученая степень

In [35]:
# формируем словарь family_status_id - family_status
family_dict = data[['family_status_id', 'family_status']]
family_dict = family_dict.drop_duplicates().reset_index(drop=True).sort_values(by='family_status_id')
family_dict

family_status_id          family_status
0                 0        женат / замужем
1                 1       гражданский брак
2                 2         вдовец / вдова
3                 3              в разводе
4                 4  не женат / не замужем

In [36]:
# Удаляем из исходного датафрейма столбцы education и family_status
data = data.drop('education',axis=1)
data = data.drop('family_status',axis=1)
data.head()

children  days_employed  dob_years  education_id  family_status_id gender  \
0       1.0           8437         42             0                 0      F   
1       1.0           4024         36             1                 0      F   
2       0.0           5623         33             1                 0      M   
3       3.0           4124         32             1                 0      M   
4       0.0           2261         53             1                 1      F   

  income_type  debt  total_income                     purpose dob_group  
0   сотрудник     0        253875               покупка жилья     40-49  
1   сотрудник     0        112080     приобретение автомобиля     30-39  
2   сотрудник     0        145885               покупка жилья     30-39  
3   сотрудник     0        267628  дополнительное образование     30-39  
4   пенсионер     0        158616             сыграть свадьбу     50-59

### Шаг 2.6. Категоризация дохода.

[*Вернуться к оглавлению*](#Оглавление)

На основании диапазонов доходов, указанных ниже, создаем столбец `total_income_category` с категориями:

    0–30000 — 'E';
    30001–50000 — 'D';
    50001–200000 — 'C';
    200001–1000000 — 'B';
    1000001 и выше — 'A'.

In [37]:
# создаем функцию, которая определяет категорию заемщика по уровню дохода
def total_income_category(value):
    try:
        if (value > 0) and (value <= 30000):
            return 'E'
        if (value > 30000) and (value <= 50000):
            return 'D'
        if (value > 50000) and (value <= 200000):
            return 'C'
        if (value > 200000) and (value <= 1000000):
            return 'B'
        if value > 1000000:
            return 'A'
        else:
            return 'Данные по доходам не могут быть меньше или раны нулю'
    except:
        return 'Данные по доходам неверные или отсутствуют'

data['total_income_category'] = data['total_income'].apply(total_income_category)
data['total_income_category'].value_counts()

C    16015
B     5042
D      350
A       25
E       22
Name: total_income_category, dtype: int64

### Шаг 2.7. Категоризация целей кредита.

[*Вернуться к оглавлению*](#Оглавление)

Создаем функцию, которая на основании данных из столбца `purpose` сформирует новый столбец `purpose_category`, в который войдут следующие категории:

    'операции с автомобилем',
    'операции с недвижимостью',
    'проведение свадьбы',
    'получение образования'.

In [38]:
# функция, которая разбивает цели на категории:
def purpose_category(val):
    if 'свадь' in val:
        return 'проведение свадьбы'
    if 'недвиж' in val:
        return 'операции с недвижимостью'
    if 'жиль' in val:
        return 'операции с недвижимостью'    
    if 'автом' in val:
        return 'операции с автомобилем'    
    if 'образов' in val:
        return 'получение образования'
    else:
        return 'цель не определена'

data['purpose_category'] = data['purpose'].apply(purpose_category)
data['purpose_category'].value_counts()
data.head()

children  days_employed  dob_years  education_id  family_status_id gender  \
0       1.0           8437         42             0                 0      F   
1       1.0           4024         36             1                 0      F   
2       0.0           5623         33             1                 0      M   
3       3.0           4124         32             1                 0      M   
4       0.0           2261         53             1                 1      F   

  income_type  debt  total_income                     purpose dob_group  \
0   сотрудник     0        253875               покупка жилья     40-49   
1   сотрудник     0        112080     приобретение автомобиля     30-39   
2   сотрудник     0        145885               покупка жилья     30-39   
3   сотрудник     0        267628  дополнительное образование     30-39   
4   пенсионер     0        158616             сыграть свадьбу     50-59   

  total_income_category          purpose_category  
0                     B  операции с недвижимостью  
1                     C    операции с автомобилем  
2                     C  операции с недвижимостью  
3                     B     получение образования  
4                     C        проведение свадьбы

## Ответы на вопросы.

### Вопрос 1: Есть ли зависимость между количеством детей и возвратом кредита в срок?

[*Вернуться к оглавлению*](#Оглавление)

In [39]:
# формируем сводную таблицу по количеству детей и задолженности:
children_debt_group = data.pivot_table(index=['children'],values='debt',aggfunc={'count', 'sum', 'mean'})

# рассчитаем вес категории количества детей по всей выборке
children_debt_group['percent_weight'] =  children_debt_group['count'] * 100 / len(data)

# выведем отсортированную таблицу по среднему значению задолженности для каждой категории
children_debt_group.sort_values(by='mean')

count      mean   sum  percent_weight
children                                       
5.0           9  0.000000     0        0.041950
0.0       14167  0.075598  1071       66.034306
3.0         330  0.081818    27        1.538175
1.0        4855  0.091658   445       22.629813
2.0        2052  0.094542   194        9.564650
4.0          41  0.097561     4        0.191107

Для анализа влияния количества детей на задолженность, исключим из выборки строки с весом менее 2%.

In [40]:
children_debt_group[children_debt_group['percent_weight']>2].sort_values(by='mean')

count      mean   sum  percent_weight
children                                       
0.0       14167  0.075598  1071       66.034306
1.0        4855  0.091658   445       22.629813
2.0        2052  0.094542   194        9.564650

**Вывод 1:**  
Выборка с количеством детей 5, 4 и 3 нерепрезентативна, т.к. составляет менее двух процентов данных.  
Зависимость между количеством детей и возвратом кредита есть. Заемщики без детей возвращают долг чаще, чем заемщики с одним и двумя детьми.


### Вопрос 2. Есть ли зависимость между семейным положением и возвратом кредита в срок?

[*Вернуться к оглавлению*](#Оглавление)

In [41]:
# формируем сводную таблицу по семейному положению и задолженности:
family_debt_group = data.pivot_table(index=['family_status_id'],values='debt',aggfunc={'count', 'sum', 'mean'})

# рассчитаем вес категории семейное положение по всей выборке
family_debt_group['percent_weight'] =  family_debt_group['count'] * 100 / len(data)

# объединим сводную таблицу с категориями семейного положения
family_debt_group = family_debt_group.merge(family_dict, on='family_status_id', how='left')

# выведем отсортированную таблицу по среднему значению задолженности для каждой категории
family_debt_group.sort_values(by='mean')

family_status_id  count      mean  sum  percent_weight  \
2                 2    959  0.065693   63        4.470029   
3                 3   1195  0.071130   85        5.570057   
0                 0  12339  0.075452  931       57.513750   
1                 1   4151  0.093471  388       19.348373   
4                 4   2810  0.097509  274       13.097791   

           family_status  
2         вдовец / вдова  
3              в разводе  
0        женат / замужем  
1       гражданский брак  
4  не женат / не замужем

**Вывод 2:**  
Зависимость между семейном положением и возвратом кредита есть. Заемщики, которые находятся или были в официальном браке возвращают кредит чаще.

### Вопрос 3. Есть ли зависимость между уровнем дохода и возвратом кредита в срок?

[*Вернуться к оглавлению*](#Оглавление)

In [42]:
# формируем сводную таблицу по доходу и задолженности:
income_debt_group = data.pivot_table(index=['total_income_category'],values='debt',aggfunc={'count', 'sum', 'mean'})

# рассчитаем вес категории дохода по всей выборке
income_debt_group['percent_weight'] =  income_debt_group['count'] * 100 / len(data)

# выведем отсортированную таблицу по среднему значению задолженности для каждой категории
income_debt_group.sort_values(by='mean')

count      mean   sum  percent_weight
total_income_category                                       
D                        350  0.060000    21        1.631397
B                       5042  0.070607   356       23.501445
A                         25  0.080000     2        0.116528
C                      16015  0.084920  1360       74.648084
E                         22  0.090909     2        0.102545

**Вывод 3:**
Выборка по доходам для заемщиков категории A (1000001 и выше), D (30001–50000) и E (0–30000) нерепрезентативна, т.к составляет менее двух процентов данных.  
Зависимость между доходом и возвратом кредита есть. Заемщики категории B (200001–1000000) возвращают кредит чаще, чем заемщики с меньшим доходом категории C (50001–200000).

### Вопрос 4. Как разные цели кредита влияют на его возврат в срок?

[*Вернуться к оглавлению*](#Оглавление)

In [43]:
# формируем сводную таблицу по цели кредита и задолженности:
purpose_debt_group = data.pivot_table(index=['purpose_category'],values='debt',aggfunc={'count', 'sum', 'mean'})

# рассчитаем вес категории дохода по всей выборке
purpose_debt_group['percent_weight'] =  purpose_debt_group['count'] * 100 / len(data)

# выведем отсортированную таблицу по среднему значению задолженности для каждой категории
purpose_debt_group.sort_values(by='mean')

count      mean  sum  percent_weight
purpose_category                                              
операции с недвижимостью  10811  0.072334  782       50.391535
проведение свадьбы         2324  0.080034  186       10.832479
получение образования      4013  0.092200  370       18.705137
операции с автомобилем     4306  0.093590  403       20.070849

**Вывод 4:**  
Кредиты с целью получения образования и на операции с автомобилем возвращают реже.  
Кредиты на операции с недвижимостью и проведение свадьбы возвращают чаще.

## Общий вывод:

[*Вернуться к оглавлению*](#Оглавление)

С целью выяснить, влияет ли семейное положение и количество детей клиента на факт погашения кредита в срок были выполнены следующие задачи:
- считана таблица с данными и изучена общая информация о данных;
- выполнена предобработка данных: заполнены пропуски, проанализированы данные на наличие аномалий, данные приведены к нужному типу, данные проверены на наличие явных и неявных дубликатов. Созданы новые датафреймы с категориями образования и семейного положения заемщиков. Создан столбец с категориями по уровню дохода заемщика. Создан столбец с категориями по целям заема.
- проведен анализ и сделаны выводы: есть ли зависимость между количеством детей и возвратом кредита в срок, есть ли зависимость между семейным положением и возвратом кредита в срок, есть ли зависимость между уровнем дохода и возвратом кредита в срок, как разные цели кредита влияют на его возврат в срок?
- сделаны вывод.

По результамт проведения анализа получены следующие выводы:
* Заемщики без детей возвращают долг чаще, чем заемщики с детьми.  
* Заемщики, которые находятся или были в официальном браке возвращают кредит чаще.
* Заемщики с высоким доходом от 200000 возвращают кредит чаще, чем заемщики с меньшим доходом.  
* Кредиты на операции с недвижимостью и проведение свадьбы возвращают чаще.
* Кредиты с целью получения образования и на операции с автомобилем возвращают реже.